### DISCUSSION TITLE
Running horizon:watch runs multiple instances instead of killing the old one

### DISCUSSION THREAD

**Discussion by RyanPaiva56** [spatie/laravel-horizon-watcher#19] [Q&A]
I absolutely love the idea of this package. When I run it though, executing one job will end up running it multiple times and my machine's processes get filled up with php instances.

> Horizon started successfully.
>    INFO  Change detected! Restarting horizon...  
> 
> Horizon started successfully.
>   2023-09-03 01:22:22 App\Jobs\Events\ProcessSessions ................ RUNNING
>   2023-09-03 01:23:52 App\Jobs\Events\ProcessSessions ................ RUNNING
>   2023-09-03 01:23:52 App\Jobs\Events\ProcessSessions .......... 139.38ms FAIL
>   2023-09-03 01:24:47 App\Jobs\Events\ProcessSessions ............ 2m 24s DONE

My processes:

![Screenshot from 2023-09-02 21-27-02](https://github.com/spatie/laravel-horizon-watcher/assets/595049/65f946bd-e408-4224-b7e7-7cd01f532ec4)

All of those processes go away after I end horizon:watch

As far as I know everything is pretty stock for both laravel and for this package, I didn't modify anything.

**Comment by freekmurze** [comment]
I does work on my configuration (got a Mac here). Feel free to submit a PR that makes this work in your environment.

**Comment by Floppy012** [comment]
I've had this running in the background while doing different tasks (programming, graphics, etc.). And suddenly everything froze. I managed to open up a terminal to find out that **334 horizon**  processes were eating up ~48 GB RAM and using up 12 Cores. I'm using Linux and have horizon running inside a docker container using sail.

This is also a problem when trying to use Xdebug (which I wasn't in the above case). PHPStorm allows for a maximum of 10 simultaneous debugger connections. Because the old horizon processes aren't terminated those connections are all blocked pretty quickly causing the whole app to freeze.

@freekmurze if you need any more info, please lmk

**Comment by nmalcolm** [comment]
Same here. I'm using Horizon watcher with Sail on Mac and after a while Laravel throws "SQLSTATE[HY000] [2002] php_network_getaddresses: getaddrinfo for mysql failed: Name or service not known" because of Horizon watcher. When I `CTRL+C` on watcher I get ~40 "Shutting down" messages. Pretty big pain point right now.

**Comment by DragosMocrii** [comment]
For anyone dealing with accumulating processes when files are changed - check that your container has the `pgrep` command available. This command is used internally by horizon. For example, it is not available by default in the FrankenPHP Debian based images, so you need to add a command like `RUN apt-get install -y procps` to get `pgrep` command available.

**Comment by mreduar** [comment]
Had `pgrep` installed and still got the pileup, so different cause for me.

`fromShellCommandline('php artisan horizon')` wraps it in a shell, so Symfony
holds the sh PID, not PHP's. `stop()` kills the sh, sh doesn't pass the signal
down, and the master is left orphaned at PPID=1 with all its workers. Watcher
starts a new one on top. Got to 62 masters / 605 processes / 23.6GB before my
VM died.

Fix:

```php
// config/horizon-watcher.php
'command' => 'exec php artisan horizon',
```
exec makes the shell become PHP under the same PID, so stop() actually
reaches it. 5 reloads now leave 1 master instead of 5. Maybe make it the default? idk
